**ЕВСЕЕВ ЕВГЕНИЙ. РИ-400024**

1. Polars

In [59]:
import polars as pl
import pandas as pd
import numpy as np


In [60]:
df=pl.read_csv("train1.csv")

In [61]:
df

shape: (891, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen ┆ … ┆ A/5 21171  ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris           ┆   ┆            ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.    ┆ … ┆ PC 17599   ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley     ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Flor…           ┆   ┆            ┆         ┆       ┆          │
│ 3           ┆ 1        ┆ 3      ┆ Heikkinen, Miss. ┆ … ┆ STON/O2.   ┆ 7.925   ┆ null  ┆ S        │
│             ┆          ┆        ┆ Laina            ┆   ┆ 3101282    ┆         ┆       ┆          │
│ 4           ┆ 1        ┆ 1      ┆ Futrelle, Mrs.   ┆ … ┆ 113803     ┆ 53.1    ┆ C123  ┆ S        │
│             ┆          ┆        ┆ Jacques Heath    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Li…             ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 888         ┆ 1        ┆ 1      ┆ Graham, Miss.    ┆ … ┆ 112053     ┆ 30.0    ┆ B42   ┆ S        │
│             ┆          ┆        ┆ Margaret Edith   ┆   ┆            ┆         ┆       ┆          │
│ 889         ┆ 0        ┆ 3      ┆ Johnston, Miss.  ┆ … ┆ W./C. 6607 ┆ 23.45   ┆ null  ┆ S        │
│             ┆          ┆        ┆ Catherine Helen  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ …                ┆   ┆            ┆         ┆       ┆          │
│ 890         ┆ 1        ┆ 1      ┆ Behr, Mr. Karl   ┆ … ┆ 111369     ┆ 30.0    ┆ C148  ┆ C        │
│             ┆          ┆        ┆ Howell           ┆   ┆            ┆         ┆       ┆          │
│ 891         ┆ 0        ┆ 3      ┆ Dooley, Mr.      ┆ … ┆ 370376     ┆ 7.75    ┆ null  ┆ Q        │
│             ┆          ┆        ┆ Patrick          ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

In [62]:
df.describe()

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ describe   ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%        ┆ 224.0       ┆ 0.0      ┆ 2.0      ┆ … ┆ null      ┆ 7.925     ┆ null  ┆ null     │
│ 50%        ┆ 446.0       ┆ 0.0      ┆ 3.0      ┆ … ┆ null      ┆ 14.4542   ┆ null  ┆ null     │
│ 75%        ┆ 669.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ null      ┆ 31.0      ┆ null  ┆ null     │
│ max        ┆ 891.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ WE/P 5735 ┆ 512.3292  ┆ T     ┆ S        │
└────────────┴─────────────┴──────────┴──────────┴───┴───────────┴───────────┴───────┴──────────┘

Количество классов

In [63]:
classes=df.groupby('Pclass').agg(pl.col('Pclass').count().alias('PassengerCount'))

<ipython-input-63-ac4ecbce5c13>:1: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  classes=df.groupby('Pclass').agg(pl.col('Pclass').count().alias('PassengerCount'))


In [64]:
print(classes)

shape: (3, 2)
┌────────┬────────────────┐
│ Pclass ┆ PassengerCount │
│ ---    ┆ ---            │
│ i64    ┆ u32            │
╞════════╪════════════════╡
│ 2      ┆ 184            │
│ 3      ┆ 491            │
│ 1      ┆ 216            │
└────────┴────────────────┘


Количество пассажиров по полу

In [65]:
malefemale = df.filter(df['Survived'] == 1).groupby('Sex').agg(pl.col('Sex').count().alias('Survived_Count'))

<ipython-input-65-585511aaff0d>:1: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  malefemale = df.filter(df['Survived'] == 1).groupby('Sex').agg(pl.col('Sex').count().alias('Survived_Count'))


In [66]:
print(malefemale)

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ Survived_Count │
│ ---    ┆ ---            │
│ str    ┆ u32            │
╞════════╪════════════════╡
│ male   ┆ 109            │
│ female ┆ 233            │
└────────┴────────────────┘


Фильтр по возрасту

In [67]:
fortyfour = df.filter(df['Age'] > 44)
print(fortyfour.head())


shape: (5, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0

2. *Pandas*

In [68]:
dfpa=pd.read_csv("train1.csv")

In [69]:
dfpa

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

Cредний возраст пассажиров и его стандартное отклонение

In [70]:
pip install bottleneck

In [71]:
import bottleneck as bn
mid = bn.nanmean(df['Age'])
otkl = bn.nanstd(df['Age'])
print(f"Средний возраст пассажиров: {mid}")
print(f"Стандартное отклонение возраста пассажиров: {otkl}")

Средний возраст пассажиров: 29.69911764705882
Стандартное отклонение возраста пассажиров: 14.516321150817317


In [72]:
dfpa['Fare_new'] = dfpa.apply(lambda row: row['Fare'] * 1.3 if row['Pclass'] in [1, 2] else row['Fare'] * 1.1, axis=1)

In [73]:
print(dfpa[['Pclass', 'Fare', 'Fare_new']].head())

   Pclass     Fare  Fare_new
0       3   7.2500   7.97500
1       1  71.2833  92.66829
2       3   7.9250   8.71750
3       1  53.1000  69.03000
4       3   8.0500   8.85500


3. *Оптимизация типов Pandas*

In [74]:
df3=pd.read_csv("Housing.csv")

In [75]:
df3

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

In [76]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [78]:
#оптимизируем данные. меняем тип object на category, также анализируем int данные и сокращаем диапазон до int8 или int16 для того, чтобы меньше памяти расходовалось

df_optimized = df3.copy()
df_optimized['price'] = df_optimized['price'].astype('int32')
df_optimized['area'] = df_optimized['area'].astype('int16')
df_optimized['bedrooms'] = df_optimized['bedrooms'].astype('int8')
df_optimized['bathrooms'] = df_optimized['bathrooms'].astype('int8')
df_optimized['stories'] = df_optimized['stories'].astype('int8')
df_optimized['parking'] = df_optimized['parking'].astype('int8')
df_optimized['mainroad'] = df_optimized['mainroad'].astype('category')
df_optimized['guestroom'] = df_optimized['guestroom'].astype('category')
df_optimized['basement'] = df_optimized['basement'].astype('category')
df_optimized['hotwaterheating'] = df_optimized['hotwaterheating'].astype('category')
df_optimized['airconditioning'] = df_optimized['airconditioning'].astype('category')
df_optimized['prefarea'] = df_optimized['prefarea'].astype('category')
df_optimized['furnishingstatus'] = df_optimized['furnishingstatus'].astype('category')







In [79]:
df_optimized.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int16   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int16(1), int32(1), int8(4)
memory usage: 10.0 KB


Поменял тип данных. Потребление памяти снизилось примерно на 45 KB. С 55.5 Кб до 10 Кб.